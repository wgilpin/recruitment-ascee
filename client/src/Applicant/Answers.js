import React from 'react';
import PropTypes from 'prop-types';
import styles from './ApplicantStyles';
import FetchData from '../common/FetchData';

const propTypes = {
  onReadyStatus: PropTypes.func,
  readonly: PropTypes.bool,
  targetId: PropTypes.number,
  answers: PropTypes.object,
  questions: PropTypes.object,
};

const defaultProps = {
  answers: {},
};

export default class Answers extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dirtyAnswers: false,
      answers: props.answers,
    };
  }

  componentDidMount() {
    this.checkAllQuestionsAnswered();
  }

  checkAllQuestionsAnswered = () => {
    const { answers, dirtyAnswers } = this.state;
    if (Object.keys(answers).length === 0 || dirtyAnswers) {
      this.props.onReadyStatus(false);
    }
    this.props.onReadyStatus (
      Object.values(answers).filter(q => q.length === 0).length === 0
    );
  };

  stateToParams = () => {
    return {
      answers: Object.entries(this.state.answers).map(
        ([question_id, text]) => ({ question_id, text })
      ),
    };
  };

  doPut = (scope, params) => new FetchData(scope).put(params);

  handleSaveAnswers = () => {
    const fetcher = this.props.fetcher || this.doPut;
    fetcher({ scope: 'answers' }, this.stateToParams()).then(() => {
      this.setState({ dirtyAnswers: false }, this.checkAllQuestionsAnswered);
      window.alert('Saved');
    });
  };

  handleAnswerChanged = e => {
    const { id, value } = e.target;
    this.setState(
      {
        dirtyAnswers: true,
        answers: {
          ...this.state.answers,
          [id]: value,
        },
      },
      this.checkAllQuestionsAnswered,
    );
  };

  render() {
    const { questions } = this.props;
    return (
      <React.Fragment>
        <h2 style={styles.heading}>Recruitment Questions</h2>
        {Object.keys(questions || {}).map(key => {
          const question = questions[key];
          const answer = this.state.answers[key];
          return (
            <div key={key}>
              <div style={styles.padded}>{question}</div>
              {this.props.readonly ? (
                <div style={styles.answerText} id={key} >
                  {answer.split('\n').map(line => <div>{line}</div>)}
                </div>
              ) : (
                <textarea
                  style={styles.answer}
                  id={key}
                  readOnly={this.props.readonly}
                  onChange={this.handleAnswerChanged}
                >
                  {answer}
                </textarea>
              )}
              <hr style={styles.hr} />
            </div>
          );
        })}
        {this.state.dirtyAnswers && !this.props.readonly && (
          <button style={styles.primaryButton} onClick={this.handleSaveAnswers}>
            Save
          </button>
        )}
      </React.Fragment>
    );
  }
}

Answers.propTypes = propTypes;
Answers.defaultProps = defaultProps;