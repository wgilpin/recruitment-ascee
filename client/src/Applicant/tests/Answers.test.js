import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Answers from '../Answers';

let mockFetch;

let serverState;

beforeEach(() => {
  window.alert = () => {};
  serverState = {
    questions: {
      '7': 'How long have you been playing Eve for?',
      '15': 'PVP or PVE? Why?',
      '16': 'Do you like this new application ?',
      '17': 'Can you write reactJs? If so, why am I writing this app??',
      '18': 'Can you confirm that you are a spai ?',
    },
    answers: {
      has_application: true,
      questions: {
        '7': {
          answer: 'app 2 5 years',
          question: 'How long have you been playing Eve for?',
          user_id: 2114725334,
        },
        '15': {
          answer: 'app 2 PVP',
          question: 'PVP or PVE? Why?',
          user_id: 2114725334,
        },
        '16': {
          answer:
            'from models import Note, Answer, User, Character, db, Application, Admin, Recruiter\nfrom security import has_applicant_access, is_admin, is_senior_recruiter, is_recruiter, \\\n    user_admin_access_check, user_application_access_check\nimport cachetools\nfrom exceptions import BadRequestException, ForbiddenException\nfrom models import Question\nfrom:',
          question: 'Do you like this new application ?',
          user_id: 2114725334,
        },
        '17': {
          answer: 'why would I?',
          question: 'Can you write reactJs? If so, why am I writing this app??',
          user_id: 2114725334,
        },
        '18': {
          answer: 'I am',
          question: 'Can you confirm that you are a spai ?',
          user_id: 2114725334,
        },
      },
    },
    characters: {
      info: {
        '94399527': {
          corporation_id: 98409330,
          corporation_name: 'Ascendance',
          name: 'Semirhage Boan',
        },
        '2114029936': {
          corporation_id: 98409330,
          corporation_name: 'Ascendance',
          name: 'The Flight Lieutenant',
        },
        '2114725334': {
          corporation_id: 98409330,
          corporation_name: 'Ascendance',
          name: 'Entropics Applicant',
        },
      },
    },
  };

  mockFetch = jest.fn().mockImplementation((scope, params) => {
    return Promise.resolve({
      info: serverState[scope.scope],
    });
  });
});

describe('<Answers>', () => {
  it('renders without crashing', () => {
    mount(
      <Answers
        fetcher={mockFetch}
        answers={serverState.answers}
        questions={serverState.questions}
        onReadyStatus={() => null}
      />
    );
  });

  it('checks when all questions have answers', () => {
    const onReadyFn = jest.fn();
    const wrapper = shallow(
      <Answers
        fetcher={mockFetch}
        answers={serverState.answers}
        questions={serverState.questions}
        onReadyStatus={onReadyFn}
      />
    )
      .instance()
      .handleSaveAnswers();
    expect(onReadyFn).toHaveBeenCalled();
  });

  
  it('matches snapshot', async () => {
    const adminConfig = mount(
      <Answers
        fetcher={mockFetch}
        answers={serverState.answers}
        questions={serverState.questions}
        onReadyStatus={() => null}
      />
    );

    const tree = renderer.create(adminConfig);
    await Promise.resolve();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
