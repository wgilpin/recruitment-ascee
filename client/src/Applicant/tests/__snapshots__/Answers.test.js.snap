// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Answers> matches snapshot 1`] = `
Array [
  <h2
    style={
      Object {
        "color": "#01799A",
      }
    }
  >
    Recruitment Questions
  </h2>,
  <div>
    <div
      style={
        Object {
          "padding": 6,
        }
      }
    >
      How long have you been playing Eve for?
    </div>
    <textarea
      id="7"
      onChange={[Function]}
      style={
        Object {
          "backgroundColor": "#222",
          "border": "none",
          "color": "white",
          "height": "60px",
          "padding": "6px",
          "width": "90%",
        }
      }
    />
    <hr
      style={
        Object {
          "borderColor": "#555555",
          "borderWidth": "1px",
        }
      }
    />
  </div>,
  <div>
    <div
      style={
        Object {
          "padding": 6,
        }
      }
    >
      PVP or PVE? Why?
    </div>
    <textarea
      id="15"
      onChange={[Function]}
      style={
        Object {
          "backgroundColor": "#222",
          "border": "none",
          "color": "white",
          "height": "60px",
          "padding": "6px",
          "width": "90%",
        }
      }
    />
    <hr
      style={
        Object {
          "borderColor": "#555555",
          "borderWidth": "1px",
        }
      }
    />
  </div>,
  <div>
    <div
      style={
        Object {
          "padding": 6,
        }
      }
    >
      Do you like this new application ?
    </div>
    <textarea
      id="16"
      onChange={[Function]}
      style={
        Object {
          "backgroundColor": "#222",
          "border": "none",
          "color": "white",
          "height": "60px",
          "padding": "6px",
          "width": "90%",
        }
      }
    />
    <hr
      style={
        Object {
          "borderColor": "#555555",
          "borderWidth": "1px",
        }
      }
    />
  </div>,
  <div>
    <div
      style={
        Object {
          "padding": 6,
        }
      }
    >
      Can you write reactJs? If so, why am I writing this app??
    </div>
    <textarea
      id="17"
      onChange={[Function]}
      style={
        Object {
          "backgroundColor": "#222",
          "border": "none",
          "color": "white",
          "height": "60px",
          "padding": "6px",
          "width": "90%",
        }
      }
    />
    <hr
      style={
        Object {
          "borderColor": "#555555",
          "borderWidth": "1px",
        }
      }
    />
  </div>,
  <div>
    <div
      style={
        Object {
          "padding": 6,
        }
      }
    >
      Can you confirm that you are a spai ?
    </div>
    <textarea
      id="18"
      onChange={[Function]}
      style={
        Object {
          "backgroundColor": "#222",
          "border": "none",
          "color": "white",
          "height": "60px",
          "padding": "6px",
          "width": "90%",
        }
      }
    />
    <hr
      style={
        Object {
          "borderColor": "#555555",
          "borderWidth": "1px",
        }
      }
    />
  </div>,
]
`;
